src/ngx_stream_echo_module.c

```diff
@@ -554,24 +554,41 @@ ngx_stream_echo_exec_echo(ngx_stream_session_t *s,
         return NGX_OK;
     }
 
-    out = ngx_chain_get_free_buf(c->pool, &ctx->free);
-    if (out == NULL) {
-        return NGX_ERROR;
-    }
 
-    out->buf->memory = 1;
+    if (c->type == SOCK_STREAM) {
+        out = ngx_chain_get_free_buf(c->pool, &ctx->free);
+        if (out == NULL) {
+            return NGX_ERROR;
+        }
 
-    out->buf->pos = cmd->data.buffer.data;
-    out->buf->last = out->buf->pos + cmd->data.buffer.len;
+        out->buf->memory = 1;
 
-    out->buf->tag = (ngx_buf_tag_t) &ngx_stream_echo_module;
+        out->buf->pos = cmd->data.buffer.data;
+        out->buf->last = out->buf->pos + cmd->data.buffer.len;
 
-    rc = ngx_chain_writer(&ctx->writer, out);
+        out->buf->tag = (ngx_buf_tag_t) &ngx_stream_echo_module;
 
-    dd("chain writer returned: %d", (int) rc);
 
-    ngx_chain_update_chains(c->pool, &ctx->free, &ctx->busy, &out,
-                            (ngx_buf_tag_t) &ngx_stream_echo_module);
+        rc = ngx_chain_writer(&ctx->writer, out);
+
+        dd("chain writer returned: %d", (int) rc);
+
+        ngx_chain_update_chains(c->pool, &ctx->free, &ctx->busy, &out,
+                                (ngx_buf_tag_t) &ngx_stream_echo_module);
+
+    }
+    else { /* SOCK_DGRAM */
+        rc = ngx_send(c, cmd->data.buffer.data, cmd->data.buffer.len);
```

Comment:
We should use `c->send()` instead of `ngx_send()` here. Fixing this should fix the "Destination address required" send() error you were seeing in in #5.